[Kernels] Migrate CPU matmul kernels from LegacyUnsafePointer to UnsafePointer

KrxGu opened the pull request:
Fixes #5670

This PR migrates the CPU matmul kernel implementations to use the new `UnsafePointer` API, removing the deprecated `LegacyUnsafePointer` alias.

### Changes
Updated 6 files in `max/kernels/src/linalg/matmul/cpu/`:
- `impl.mojo` - Core matmul orchestration
- `apple_accelerate.mojo` - Apple Accelerate BLAS wrapper
- `default.mojo` - Default CPU microkernel
- `i8mm.mojo` - ARM i8mm instruction microkernel
- `neon.mojo` - ARM NEON microkernel
- `vnni.mojo` - x86 VNNI microkernel

### Implementation Notes
The new `UnsafePointer` API requires explicit mutability (`mut=True/False`) and origin tracking. Where needed for compatibility with unmigrated code (like `_Accumulator`) and FFI interfaces, we use `rebind[]` to convert between pointer types.

### Testing
- All matmul tests pass
- Code formatted with `bazelw run //bazel/lint:format`

KrxGu committed:
Migrate CPU matmul kernels from LegacyUnsafePointer to UnsafePointer

Updated 6 CPU matmul kernel files to use the new UnsafePointer API with
explicit mutability and origin tracking. The migration uses interop with
LegacyUnsafePointer where needed for compatibility with unmigrated code
like _Accumulator and FFI interfaces.

Fixes #5670

KrxGu committed:
Remove unnecessary pointer conversions in i8mm call sites

The c_ptr from LayoutTensor is already UnsafePointer, so no need to
convert to LegacyUnsafePointer at the call sites. The conversion happens
inside _store_c_tile where it's actually needed for partial_simd_store.

NathanSWard commented on max/kernels/src/linalg/matmul/cpu/i8mm.mojo:
```diff
@@ -102,12 +102,16 @@ struct LoadStore_i8mm[
     fn _store_c_tile(
         mut self,
-        c_ptr: UnsafePointer[Scalar[Self.dtype]],
+        c_ptr: UnsafePointer[mut=True, Scalar[Self.dtype], **_],
```
```suggestion
        c_ptr: UnsafePointer[mut=True, Scalar[Self.dtype]],
```

No need to add the `**_` syntax here as the previous pointer had it's address space bound to `GENERIC`.

KrxGu replied:
Thanks for the review! I tried removing the **_ syntax, but ran into type inference issues. The functions receive pointers from LayoutTensor[mut=True, **_] which has a generic origin, and without **_ the compiler can't infer the origin parameter at call sites. 

I kept the **_ syntax for now - let me know if there's a better pattern I should use!

NathanSWard replied:
Ahh intersting, thanks for pointing that out!

soraros replied:
Is this fix, which relaxes the API as a side effect, really correct? Shouldn't we instead origin-cast the pointer obtained from the LayoutTensor?

NathanSWard commented on max/kernels/src/linalg/matmul/cpu/i8mm.mojo:
```diff
@@ -102,12 +102,16 @@ struct LoadStore_i8mm[
     ):
-        var c_ptr_loc = c_ptr.offset(tile_n_idx)
+        # Convert to LegacyUnsafePointer for partial_simd_store compatibility
+        var c_ptr_legacy = rebind[LegacyUnsafePointer[Scalar[Self.dtype]]](
```
**question**
Do we have to convert this to a `LegacyUnsafePointer`? Even though `partial_simd_store` takes a `LegacyUnsafePointer`, you _should_ be able to pass an `UnsafePointer` to it without any needed explicit conversions.

NathanSWard commented on max/kernels/src/linalg/matmul/cpu/impl.mojo:
```diff
@@ -388,7 +388,7 @@ struct TiledMatmul[
     fn _view_buffer_as(
         self,
-        b_packed_ptr: UnsafePointer[Scalar[Self.b_type]],
+        b_packed_ptr: UnsafePointer[Scalar[Self.b_type], **_],
```
```suggestion
        b_packed_ptr: UnsafePointer[Scalar[Self.b_type]],
```
Ditto here, no need to unbind the `address_space` parameter :)

NathanSWard commented on max/kernels/src/linalg/matmul/cpu/impl.mojo:
```diff
@@ -595,9 +595,9 @@ fn _matmul_cpu_impl[
-            a_packed_ptr = UnsafePointer[Scalar[a.type]].alloc(
+            a_packed_ptr = LegacyUnsafePointer[Scalar[a.type]].alloc(
                 mh * kh, alignment=alignment
             )
```
Ideally these should also be updated to use the new `UnsafePointer` and `alloc` methods.
```mojo
var a_packed_ptr = UnsafePointer[Scalar[a.type], MutOrigin.external]()
if use_i8mm:
    a_packed_ptr = alloc[Scalar[a.type]](mh * kh, alignment=alignment)
```

NathanSWard commented on max/kernels/src/linalg/matmul/cpu/neon.mojo:
```diff
@@ -141,7 +141,7 @@ struct Inner_matmul_neon(InnerMatmulKernel, Movable):
             else:
                 acc.load(
-                    rebind[UnsafePointer[Scalar[c.dtype]]](c_ptr),
+                    rebind[LegacyUnsafePointer[Scalar[c.dtype]]](c_ptr),
```
**questions**
Just double checking that the `rebind` is necessary here and we can't just pass `c_ptr` directly?

KrxGu replied:
The rebind is necessary here because _Accumulator in accumulate.mojo still uses `from memory import LegacyUnsafePointer as UnsafePointer`. The load() and store() methods expect LegacyUnsafePointer, so we need to convert from the new UnsafePointer.

If _Accumulator gets migrated to the new API in the future, we can remove these rebind calls.

soraros replied:
I thought we have two-way implicit conversion between the two types?

NathanSWard commented on max/kernels/src/linalg/matmul/cpu/apple_accelerate.mojo:
```diff
@@ -261,14 +260,16 @@ fn apple_gemv[
-        transposed_b_ptr = UnsafePointer[Scalar[b.type]].alloc(b.num_elements())
+        transposed_b_ptr = LegacyUnsafePointer[Scalar[b.type]].alloc(
+            b.num_elements()
+        )
```
Ditto using `UnsafePointer` and the free standing `alloc` function.

KrxGu committed:
Address review feedback: use new alloc API and remove unnecessary conversions

- Use free-standing alloc() instead of LegacyUnsafePointer.alloc()
- Remove LegacyUnsafePointer conversion in i8mm _store_c_tile
- Keep **_ origin syntax for proper type inference
- Keep rebind for neon.mojo since _Accumulator uses LegacyUnsafePointer
